### PR TITLE
added neural ode tutorial :books:

### PR DESCRIPTION
**Pull request by shivance**
# Pull Request Template

## Description
Added tutorial for using deepchem with torchdiffeq
Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
This is a good start! To get this tutorial closer to ready to merge we need to add in a lot more descriptive text. Here are some topics to cover in the text:

1. What is a neural ODE? When do we want to use one?
2. What is torchdiffeq? What is its API
3. What is our goal in the tutorial

In general, each block of code should be preceded by a paragraph (or sentence at least) explaining what that code is going to do. Think of this as writing a report or book chapter

**Comment by shivance** [comment]
@rbharath I've done the requested changes. Please let me know if there is scope of further improvement.

File: examples/tutorials/Using_Torchdiffeq_in_Deepchem.ipynb [hunk 1]
```diff
@@ -0,0 +1,497 @@
+{
```
[the remark this answers is not in the document]
Comment (shivance) [reply examples/tutorials/Using_Torchdiffeq_in_Deepchem.ipynb:1]: Done @rbharath

**Comment by peastman** [comment]
This is a nice start!  The main thing that I think could be improved is to better explain the concepts, especially regarding training.  Here are the major concepts as I see them.

- We are dealing with ODEs of the form dz/dt = f(z, t).  You mostly explain that well.  It isn't clear whether that's actually a restriction, though, or whether this technique can also be used for higher order ODEs or ones where f() can depend on derivatives of z.
- Give f(z, t) and the value of z for some particular t, you can integrate the equation to find z at other values of t.  You explain this well.
- In some cases you don't know what f(z, t) is, so instead you want to learn it from data.  That is what neural ODEs are all about.  You mostly explain this well, but it could be a little clearer.

But how do we learn it?  What kind of training data do we need?  If we had a series of z(t) curves we wanted to reproduce, that would make sense.  But instead you load... a set of molecules?  What do they have to do with learning ODEs?  What does z represent in this case?  What about t?  What does f(z, t) represent?  How does the training algorithm work?  What loss function are we minimizing?  The connection between the example and all the discussion that has come before is not clear.

**Comment by shivance** [comment]
@rbharath @peastman How about splitting this notebook into two ? 
Covering theory + normal usage in one notebook 
Using torchdiffeq with deepchem torchmodel and pre-built modules in second?

**Comment by rbharath** [comment]
I think two notebooks could be reasonable here. You might also consider moving the notebooks into series of tutorials on differential equations (where the PINNs tutorials currently reside)

**Comment by peastman** [comment]
I think either way is fine.  There's nothing wrong with a single long notebook.  But if you think splitting it would be better, that's fine too.  It depends on whether the first notebook can stand on its own as a useful tutorial.  If all it does is discuss theory that will be used in the second notebook, it probably makes more sense to keep them as one notebook.

**Comment by shivance** [comment]
@rbharath regarding the `normalizing flow(nf)` which you asked to mention in the tutorial somewhere as you felt neural odes somehow point to nf . Check this out [1](https://slideslive.com/38917901/neural-ordinary-differential-equations-for-continuous-normalizing-flows?locale=cs) and [2](https://arxiv.org/pdf/2006.00104.pdf).

**Comment by rbharath** [comment]
This is a good improvement! Still a few typos but other than that looks good on my end.

**Comment by shivance** [comment]
Hey @rbharath fixed that. Typos are so annoying 😵‍💫

**Comment by shivance** [comment]
@rbharath I think it's ready to be merged.

**Comment by shivance** [comment]
@rbharath Please review

File: examples/tutorials/About_nODE_Using_Torchdiffeq_in_Deepchem.ipynb [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+      "source": [
+        "# What is Neural Differential Equation ?\n",
+        "\n",
+        "A neural differential equation is a differential equation using a neural network to parameterize the vector field. The canonical example is a neural ordinary differential equation :\n",
+        "\n",
+        "$y(0) = y_0$\n",
+        "\n",
+        "$\\frac{dy}{dt} (t) = f_\\theta(t,y(t)) $\n",
+        "\n",
+        "Here θ represents some vector of learnt parameters, $ f_\\theta : \\mathbb{R} \\times \\mathbb{R}^{d_1 \\times ... \\times d_k}$ is any standard neural architecture and $ y:[0, T] → \\mathbb{R}^{d_1 \\times ... d_k} $ is the solution. For many applications $f_\\theta$ will just be a simple feedforward network. Here $d_i $ is the dimension. \n",
```
Comment (rbharath) [inline examples/tutorials/About_nODE_Using_Torchdiffeq_in_Deepchem.ipynb:251]: As a reminder, make sure to add on an explanation of what d_1, ..., d_k are here.

File: examples/tutorials/README.md [hunk 3]
```diff
@@ -91,3 +94,5 @@ tutorials discuss about using DeepChem for specific applications.
 
 ### Physics Informed Neural Networks
```
Comment (rbharath) [inline examples/tutorials/README.md:95]: This title should be changed. How about

### Deep Differential Equations

**Review by rbharath — COMMENTED** [review]
A couple of minor comments

File: examples/tutorials/About_nODE_Using_Torchdiffeq_in_Deepchem.ipynb (hunk 2 above)
Comment (shivance) [reply examples/tutorials/About_nODE_Using_Torchdiffeq_in_Deepchem.ipynb:251]: In last line in cell I've already added no, `Here $d_i $ is the dimension. \n"`

**Comment by shivance** [comment]
@rbharath

**Review by rbharath — APPROVED** [review]
LGTM!
